lib/rex/crypto/chacha20.rb

```diff
@@ -0,0 +1,23 @@
+# -*- coding: binary -*-
+
+module Rex
+  module Crypto
+    def self.chacha_encrypt(key, iv, plaintext)
+      cipher = OpenSSL::Cipher.new('chacha20')
+      cipher.encrypt
+      cipher.key = key
+      cipher.iv = iv
```

Comment:
This appears to reset the IV every time there is an encrypt or decrypt. Doing so means that the same plaintext on the wire will yield the same ciphertext on the wire for the duration of the session. For instance, if someone types 'dir', it will show up as (for example) 'ada' every time, making wire-analysis much easier. This is similar to 'ECB' https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Electronic_Codebook_(ECB) modes in block ciphers. To encrypt the stream without this issue, you need to save the crypto context for encrypt/decrypt for the duration of the session without resetting the IV. I see a instance-variable IV in the code that calls this function, but it's not clear if it's mutated here and passed back to the caller.

Reply:
I modified the chacha implementation to keep track of the context throughout the session, and while testing, it looks like the same command results in unique ciphertext each time. Please let me know if the new version is a good fix!

Reply:
Nice, thanks!